clients/src/main/java/org/apache/kafka/common/record/MemoryRecords.java

```diff
@@ -679,12 +678,13 @@ public static MemoryRecords withRecords(byte magic, long initialOffset, Compress
         long logAppendTime = RecordBatch.NO_TIMESTAMP;
         if (timestampType == TimestampType.LOG_APPEND_TIME)
             logAppendTime = System.currentTimeMillis();
-        MemoryRecordsBuilder builder = new MemoryRecordsBuilder(bufferStream, magic, compressionType, timestampType,
+        try (final MemoryRecordsBuilder builder = new MemoryRecordsBuilder(bufferStream, magic, compressionType, timestampType,
```

Comment:
Note to reviewer(s):

`builder.build()` calls the `MemoryRecordsBuilder#close()` method, hence, this is not a leak in happy cases. But it will not call `close()` in case exception occurs at `builder.append()`, hence, MemoryRecordsBuilder will not be closed when an exception is thrown.